### PR TITLE
quiche: enable qlog output

### PR DESCRIPTION
**Pull request by LPardue**
NB: I hate string handing in C, so I've gone with a simple stupid option that illustrates what needs to be done to make qlog work. Please suggest better code that matches curl's preferred way of things, and take note of the TODOs 

quiche has the potential to log qlog files. To enable this, you must
build quiche with the qlog feature enabled
`cargo build --features qlog`. curl can then pass a file descriptor
to quiche, which will take ownership of the file. The FD transfer
only works on UNIX.

The convention is to enable logging when the QLOGDIR environment is
set. This should be a path to a folder where files are written with
the naming template <SCID>.qlog.

### CODE REVIEW

File: lib/vquic/quiche.c [hunk 1]
```diff
@@ -199,6 +203,24 @@ CURLcode Curl_quic_connect(struct connectdata *conn, curl_socket_t sockfd,
     return CURLE_OUT_OF_MEMORY;
   }
 
+/* TODO: detect UNIX nicely */
+#ifndef _WIN32
```
Comment (emilengler) [inline lib/vquic/quiche.c:207]: How can you be sure that every non Windows system is Unix?  I would suggest using `__unix__` if possible

File: lib/vquic/quiche.c [hunk 2]
```diff
@@ -64,6 +64,7 @@ static CURLcode http_request(struct connectdata *conn, const void *mem,
 static Curl_recv h3_stream_recv;
 static Curl_send h3_stream_send;
 
+static FILE *qlog_file; /* not thread-safe */
```
Comment (emilengler) [inline lib/vquic/quiche.c:67]: Why does it need to be a global variable

File: lib/vquic/quiche.c [hunk 3]
```diff
@@ -152,6 +153,9 @@ CURLcode Curl_quic_connect(struct connectdata *conn, curl_socket_t sockfd,
   CURLcode result;
   struct quicsocket *qs = &conn->hequic[sockindex];
   struct Curl_easy *data = conn->data;
+  const char *qlog_dir;
+  char *qlog_filename;
+  char *p;
```
Comment (emilengler) [inline lib/vquic/quiche.c:158]: Maybe you can put these definitions in their own scope below

Comment (emilengler) [reply lib/vquic/quiche.c:158]: Also why declare the variables if they are not being used on a windows system

**Comment by LPardue** [comment]
Thanks! As you might be able to tell from the PR, all quiche needs is a FD. So really the code review is hood to focus on C. I made it very quickly without consulting curl best practice etc - if someone wants to jump on and bring this up to snuff I'm not precious with sharing the glory ;)

File: lib/vquic/quiche.c (hunk 1 above)
Comment (bagder) [reply lib/vquic/quiche.c:207]: I don't. Windows is in most cases the special-case, not the other way around. And I think the check should be for `WIN32` as that's what we do all over.

File: lib/vquic/quiche.c [hunk 4]
```diff
@@ -199,6 +203,24 @@ CURLcode Curl_quic_connect(struct connectdata *conn, curl_socket_t sockfd,
...
+#ifndef _WIN32
+  qlog_dir = getenv("QLOGDIR");
+  if(qlog_dir) {
+    qlog_filename = malloc(strlen(qlog_dir) + 1 + sizeof(qs->scid) + 5);
+    /* TODO: better string concatentation and actually use the hex-formatted SCID */
+    p = qlog_filename;
+    strcpy(p, qlog_dir);
+    strcat(p, "/");
+    strcat(p, "12345678");
+    strcat(p, ".qlog");
```
Comment (bagder) [inline lib/vquic/quiche.c:216]: This could be made into a single `aprintf()` call that returns the alloced new string.

File: lib/vquic/quiche.c [hunk 5]
```diff
@@ -199,6 +203,24 @@ CURLcode Curl_quic_connect(struct connectdata *conn, curl_socket_t sockfd,
     return CURLE_OUT_OF_MEMORY;
   }
 
+/* TODO: detect UNIX nicely */
+#ifndef _WIN32
+  qlog_dir = getenv("QLOGDIR");
+  if(qlog_dir) {
+    qlog_filename = malloc(strlen(qlog_dir) + 1 + sizeof(qs->scid) + 5);
```
Comment (bagder) [inline lib/vquic/quiche.c:210]: This needs to check for malloc failure.

**Comment by LPardue** [comment]
@bagder you reminded that building without the qlog feature enabled will cause a link error, so we probably want to make the whole block conditionally compiled based on the feature. I don't know how best to do that in curl. 

In the nginx integration, we have a `--with-quiche-opt` configure flag that we can pass in things like `--features foo`. And so my candidate solution there relies are detecting the qlog feature and setting a define.

**Comment by bagder** [comment]
> if someone wants to jump on and bring this up to snuff I'm not precious with sharing the glory

Sure, I can polish it up! You've showed how to use the API very nicely.

> without the qlog feature enabled will cause a link error, so we probably want to make the whole block conditionally compiled based on the feature

Do you think it would be possible to have the functions just do nothing (but exist) if the feature isn't there? It would simplify for a user like us. We can certainly add a build-time check for the functions, but then that's a conditional build and you couldn't just replace quiche to get qlog going...

**Comment by LPardue** [comment]
> Do you think it would be possible to have the functions just do nothing (but exist) if the feature isn't there?

That's a good question for @ghedo

**Comment by bagder** [comment]
Closed by #5341

**Pull request closed without merging**